nailgun/entity_mixins.py

```diff
@@ -931,3 +937,325 @@ def update(self, fields=None):
+
+    The methods provided by this class work together. The call tree looks like
+    this::
+
+        search
+        ├── search_json
+        │   └── search_raw
+        │       └── search_payload
+        ├── search_normalize
+        └── search_filter
+
+    In short, here is what the methods do:
+
+    :meth:`search_payload`
+        Assemble a search query that can be encoded and sent to the server.
+    :meth:`search_raw`
+        Make an HTTP GET request to the server, including the payload.
+    :meth:`search_json`
+        Check the server's response for errors and decode the response.
+    :meth:`search_normalize`
+        Normalize search results so they can be used to create new entities.
+    :meth:`search`
+        Create one or more :class:`nailgun.entity_mixins.Entity` objects
+        representing the found entities and populate their fields.
+    :meth:`search_filter`
+        Read all ``entities`` and locally filter them.
+
+    See the individual methods for more detailed information.
+
+    """
+
+    def search_payload(self, fields=None, query=None):
+        """Create a search query.
+
+        Do the following:
+
+        1. Generate a search query. If ``fields`` is specified, all named
+           fields are included in the query. Otherwise, all populated fields
+           are included.
+        2. Merge ``query`` in to the generated search query.
+        3. Return the result.
+
+        The rules for generating a search query are simple, and can be
+        illustrated by example::
+
+            >>> SomeEntity(
+            ...     id=1,
+            ...     one_to_one=2,
+            ...     one_to_many=[3, 4],
+            ... ).search_payload()
+            {'id': 1, 'one_id': 2, 'many_ids': [3, 4]},
+
+        .. WARNING:: This method currently generates an extremely naive search
+            query that will be wrong in many cases.
+
+        :param fields: See :meth:`search`.
+        :param query: See :meth:`search`.
+        :returns: A dict that can be encoded as JSON and used in a search.
+
+        """
+        if fields is None:
+            fields = set(self.get_values().keys())
+        if query is None:
+            query = {}
+
+        payload = {}
+        fields_dict = self.get_fields()
+        for field in fields:
+            value = getattr(self, field)
+            if isinstance(fields_dict[field], OneToOneField):
+                payload[field + '_id'] = value.id
+            elif isinstance(fields_dict[field], OneToManyField):
+                payload[field + '_ids'] = [entity.id for entity in value]
+            else:
+                payload[field] = value
+        payload.update(query)
+        return payload
+
+    def search_raw(self, fields=None, query=None):
+        """Search for entities.
+
+        Make an HTTP GET call to ``self.path('base')``. Return the response.
+
+        .. WARNING:: Subclasses should not fiddle with the ``fields`` or
```

Comment:
But we have similar idea for, let's say, for read method:
```
class Domain(Entity, EntityReadMixin):

    def read(self, entity=None, attrs=None, ignore=()):
        if attrs is None:
            attrs = self.read_json()
        attrs['domain_parameters_attributes'] = attrs.pop('parameters')
        return super(Domain, self).read(entity, attrs, ignore)
```
Here, we change parameter and send it to super class

Or it is completely different?